alacritty_terminal/src/ansi.rs

```diff
@@ -24,63 +24,59 @@ use vte;
 
 use crate::term::color::Rgb;
 
+fn is_rgb_separator(ch: Option<char>) -> bool {
+    ch == None || ch == Some('/') || ch == Some('\x07')
+}
+
 // Parse color arguments
-//
-// Expect that color argument looks like "rgb:xx/xx/xx" or "#xxxxxx"
+// Expect the color argument in the form "rgb:x/x/x" or "#xxx", where x is 1-4 digit hex numbers
 fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
-    let mut iter = color.iter();
-
-    macro_rules! next {
-        () => {
-            iter.next().map(|v| *v as char)
-        };
+    let len = color.len();
+    // Legacy colors are in the form "#xxx"
+    // The use of this format is no longer encouraged by xparsecolor
+    let mut is_legacy = false;
+
+    if len >= 5 && color[0] == b'#' {
+        is_legacy = true;
+    } else if len < 10 || &color[..4] != b"rgb:" {
+        return None;
     }
 
-    macro_rules! parse_hex {
-        () => {{
-            let mut digit: u8 = 0;
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit = value as u8;
-            }
+    let mut iter = color.iter().map(|v| *v as char).skip(if is_legacy { 1 } else { 4 });
+    let mut colors = [0; 3];
 
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit <<= 4;
-                digit += value as u8;
-            }
-            digit
-        }};
-    }
+    for color in &mut colors {
```

Comment:
I think we can simplify this whole block a bit.

Instead of having a lot of branching statements what should be done depending on `i` and the different modes we're currently in, I think it might be useful to just pull as many characters as each color is long and then parse them together in bulk.

So we'd first collect all elements into a string, str, [u8; X], [char; X] or similar. This can be done for legacy mode by dividing the length by three or in the new mode by taking from the iterator while `!is_separator` is true.

Now that we have that, instead of going for branching we can just use `std` and some math to calculate the correct value.

I've got a playground link that should help with how this can be done, this should be simple to adapt depending on the different input types one might choose as most reasonable (I think for us, `str` is probably best):
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fff750f405a6b7e1578ce80916163dd5